pkg/bus/bus.go

```diff
@@ -155,13 +162,68 @@ func (b *InProcBus) Dispatch(msg Msg) error {
 	return err.(error)
 }
 
+// PublishCtx function publish a message to the bus listener.
+func (b *InProcBus) PublishCtx(ctx context.Context, msg Msg) error {
```

Comment:
Since there in theory could be a listener registered using `AddListener` and one registered using `AddListenerCtx` for the same message think we need to handle that. Basically calling both any non-ctx listeners and any ctx listeners. Need to do handle this for both `Publish` and `PublishCtx` which definitely complicates the code a bit.

Reply:
I adjusted PublishCtx according to your comment, let me know if I am understood you correctly

Reply:
That won't work if you have a mixed set of listeners (with ctx and without ctx). I'm thinking something like this should work:

```go
	if listeners, exists := b.listenersWithCtx[msgName]; exists {
		params = append(params, reflect.ValueOf(ctx))

		if setting.Env == setting.Dev {
			b.logger.Warn("PublishCtx called with message listener registered using AddEventListener and should be changed to use AddEventListenerCtx", "msgName", msgName)
		}

		for _, listenerHandler := range listeners {
			// 
		}
	}
	
	// similar logic as above for b.listeners	
```